lib/datasource/maven/index.ts

```diff
@@ -136,6 +138,85 @@ function extractVersions(metadata: XmlDocument): string[] {
   return elements.map(el => el.val);
 }
 
+async function getVersionsFromMetadata(
+  dependency: MavenDependency,
+  repoUrl: string
+): Promise<string[] | null> {
+  const metadataUrl = getMavenUrl(dependency, repoUrl, 'maven-metadata.xml');
+  if (!metadataUrl) return null;
+
+  const cacheNamespace = 'datasource-maven-metadata';
+  const cacheKey = metadataUrl.toString();
+  const cachedVersions = await renovateCache.get<string[]>(
+    cacheNamespace,
+    cacheKey
+  );
+  /* istanbul ignore if */
+  if (cachedVersions) return cachedVersions;
+
+  const mavenMetadata = await downloadMavenXml(metadataUrl);
+  if (!mavenMetadata) return null;
+
+  const versions = extractVersions(mavenMetadata);
+  await renovateCache.set<string[]>(cacheNamespace, cacheKey, versions, 10);
+  return versions;
+}
+
+type ArtifactsInfo = Record<string, boolean | null>;
+
+function isValidArtifactsInfo(
+  info: ArtifactsInfo | null,
+  versions: string[]
+): boolean {
+  if (!info) return false;
+  return versions.every(v => info[v] !== undefined);
+}
+
+async function filterMissingArtifacts(
+  dependency: MavenDependency,
+  repoUrl: string,
+  versions: string[]
+): Promise<string[]> {
+  const cacheNamespace = 'datasource-maven-metadata';
+  const cacheKey = dependency.dependencyUrl;
+  let artifactsInfo: ArtifactsInfo | null = await renovateCache.get<
+    ArtifactsInfo
+  >(cacheNamespace, cacheKey);
+
+  if (!isValidArtifactsInfo(artifactsInfo, versions)) {
+    artifactsInfo = {};
+    for (const version of versions) {
+      const artifactUrl = getMavenUrl(
+        dependency,
+        repoUrl,
+        `${version}/${dependency.name}-${version}.jar`
+      );
+      if (artifactUrl) {
+        const proto = artifactUrl.protocol;
+        if (proto === 'http:' || proto === 'https:') {
+          artifactsInfo[version] = await isHttpResourceExists(artifactUrl);
+        } else {
+          artifactsInfo[version] = true;
+        }
+      }
+    }
+
+    // Retry earlier for status other than 404
```

Comment:
This is a good idea.. wondering if 6 hours is too long though. We should only attempt these URLs if the XML returns a 200 OK, so what would be the chances of a non-404 in such cases?

Reply:
For Maven, it looks like something extraordinary to get 50x errors, but for other repositories it can be produced more frequently. Does one hour seem optimal?

Reply:
Sure. It’s hopefully unlikely to get anything other than a 200 or 404 for jars if got a 200 for the index.